File: algorithms/16_kotlin/searching/BinarySearch.kt
fun binarySearch(arr: IntArray, target: Int): Int {
    var lo = 0
    var hi = arr.size - 1
    while (lo <= hi) {
        val mid = (lo + hi) / 2
        when {
            arr[mid] == target -> return mid
            arr[mid] < target -> lo = mid + 1
            else -> hi = mid - 1
        }
    }
    return -1
}

fun main() {
    val arr = intArrayOf(1, 3, 5, 7, 9, 11, 13)
    check(binarySearch(arr, 7) == 3)
    check(binarySearch(arr, 8) == -1)
    check(binarySearch(arr, 1) == 0)
    println("[Kotlin BinarySearch] Binary search verified")
}

File: algorithms/06_javascript/searching/binary_search.js
function lowerBound(arr, target) {
  let lo = 0, hi = arr.length;
  while (lo < hi) {
    const mid = (lo + hi) >> 1;
    if (arr[mid] < target) lo = mid + 1;
    else hi = mid;
  }
  return lo;
}

function upperBound(arr, target) {
  let lo = 0, hi = arr.length;
  while (lo < hi) {
    const mid = (lo + hi) >> 1;
    if (arr[mid] <= target) lo = mid + 1;
    else hi = mid;
  }
  return lo;
}

module.exports = { lowerBound, upperBound };

if (require.main === module) {
  const arr = [1, 3, 3, 3, 5, 7, 9];
  if (lowerBound(arr, 3) !== 1) throw new Error("lowerBound failed");
  if (upperBound(arr, 3) !== 4) throw new Error("upperBound failed");
  if (lowerBound(arr, 4) !== 4) throw new Error("lowerBound insertion failed");
  console.log("[JavaScript Binary Search] lower_bound / upper_bound verified");
}
